File: problem_set_2/valid_parentheses.js
function main() {
    printIsValid("[{()}]");     // Output: Input: [{()}] Output: true
    printIsValid("()[]{}");     // Output: Input: ()[]{} Output: true
    printIsValid("{}([){}]");   // Output: Input: {}([){}] Output: false
    printIsValid("");           // Output: Input:  Output: true (empty string)
    printIsValid("}");          // Output: Input: } Output: false
    printIsValid("(");          // Output: Input: ( Output: false
}

function printIsValid(brackets) {
    console.log(`Input: ${brackets}`);
    console.log(`Output: ${isValid(brackets)}`);
    console.log();
}

function isValid(brackets) {
    let stack = [];
    
    for (let i = 0; i < brackets.length; i++) {
        let bracket = brackets[i];
        
        if (bracket === '(' || bracket === '{' || bracket === '[') {
            stack.push(bracket); // Push onto stack
            // console.log(`Pushed onto stack: ${bracket}, new stack: ${stack.join(', ')}`);
        } else if (bracket === ')' || bracket === '}' || bracket === ']') {
            if (stack.length === 0) {
                // console.log(`Executed - No matching opening bracket`);
                return false; // No matching opening bracket
            }
            
            let topChar = stack.pop(); // Pop from stack
            // console.log(`Popped from stack: ${topChar}, new stack: ${stack.join(', ')}`);
            
            if ((bracket === ')' && topChar !== '(') ||
                (bracket === '}' && topChar !== '{') ||
                (bracket === ']' && topChar !== '[')) {
                // console.log(`Executed - Mismatched brackets`);
                return false; // Mismatched brackets
            }
        }
    }
    
    return stack.length === 0; // Stack should be empty if all brackets matched
}

main();

File: problem_set_2/valid_parentheses.go
package main
import "fmt"

func main() {
	printIsValid("[{()}]")     // Output: Input: [{()}] Output: true
	printIsValid("()[]{}")     // Output: Input: ()[]{} Output: true
	printIsValid("{}([){}]")   // Output: Input: {}([){}] Output: false
	printIsValid("")           // Output: Input:  Output: true (empty string)
	printIsValid("}")          // Output: Input: } Output: false
	printIsValid("(")          // Output: Input: ( Output: false
}

func printIsValid(brackets string) {
	fmt.Printf("Input: %s\n", brackets)
	fmt.Printf("Output: %v\n\n", isValid(brackets))
}

func isValid(brackets string) bool {
	stack := make([]rune, 0) // Using a slice to simulate stack of runes (characters)

	for _, bracket := range brackets {
		if bracket == '(' || bracket == '{' || bracket == '[' {
			stack = append(stack, bracket) // Push onto stack
			// fmt.Printf("Pushed onto stack: %c, new stack: %s\n", bracket, string(stack))
		} else if bracket == ')' || bracket == '}' || bracket == ']' {
			if len(stack) == 0 {
				// fmt.Println("Executed - No matching opening bracket")
				return false // No matching opening bracket
			}

			topChar := stack[len(stack)-1] // Peek top of stack
			stack = stack[:len(stack)-1]   // Pop from stack
			// fmt.Printf("Popped from stack: %c, new stack: %s\n", topChar, string(stack))

			if (bracket == ')' && topChar != '(') ||
				(bracket == '}' && topChar != '{') ||
				(bracket == ']' && topChar != '[') {
				// fmt.Println("Executed - Mismatched brackets")
				return false // Mismatched brackets
			}
		}
	}

	return len(stack) == 0 // Stack should be empty if all brackets matched
}